crates/chia-bls/src/cached_bls.rs

```diff
@@ -0,0 +1,291 @@
+// This cache is a bit weird because it's trying to account for validating
+// mempool signatures versus block signatures. When validating block signatures,
+// there's not much point in caching the pairings because we're probably not going
+// to see them again unless there's a reorg. However, a spend in the mempool
+// is likely to reappear in a block later, so we can save having to do the pairing
+// again. So caching is primarily useful when synced and monitoring the mempool in real-time.
+
+use crate::aggregate_verify as agg_ver;
+use crate::gtelement::GTElement;
+use crate::hash_to_g2;
+use crate::PublicKey;
+use crate::Signature;
+use lru::LruCache;
+use sha2::{Digest, Sha256};
+use std::collections::HashMap;
+use std::num::NonZeroUsize;
+
+#[cfg(feature = "py-bindings")]
+use pyo3::types::{PyBool, PyInt, PyList};
+#[cfg(feature = "py-bindings")]
+use pyo3::{pyclass, pymethods, PyResult};
+
+#[cfg_attr(feature = "py-bindings", pyclass(name = "BLSCache"))]
+pub struct BLSCache {
+    cache: LruCache<[u8; 32], GTElement>,  // LRUCache of hash(pubkey + message) -> GTElement
+}
+
+impl Default for BLSCache {
+    fn default() -> Self {
+        Self::new(50000)
+    }
+}
+
+impl BLSCache {
+    pub fn new(cache_size: usize) -> BLSCache {
+        let cache: LruCache<[u8; 32], GTElement> =
+            LruCache::new(NonZeroUsize::new(cache_size).unwrap());
+        Self { cache }
+    }
+
+    pub fn generator(cache_size: Option<usize>) -> Self {
+        let cache: LruCache<[u8; 32], GTElement> =
+            LruCache::new(NonZeroUsize::new(cache_size.unwrap_or(50000)).unwrap());
+        Self { cache }
+    }
+
+    fn get_pairings<M: AsRef<[u8]>>(
+        &mut self,
+        pks: &[[u8; 48]],
+        msgs: &[M],
+        force_cache: bool,
+    ) -> Vec<GTElement> {
+        let mut pairings: Vec<Option<GTElement>> = vec![];
+        let mut missing_count: usize = 0;
+
+        for (pk, msg) in pks.iter().zip(msgs.iter()) {
+            let mut hasher = Sha256::new();
+            hasher.update(pk);
+            hasher.update(msg); // pk + msg
+            let h: [u8; 32] = hasher.finalize().into();
+
+            let pairing = self.cache.get(&h).cloned();
+
+            if !force_cache && pairing.is_some() {
+                // Heuristic to avoid more expensive sig validation with pairing
+                // cache when it's empty and cached pairings won't be useful later
+                // (e.g. while syncing)
+                missing_count += 1;
+                if missing_count > pks.len() / 2 {
```

Comment:
I think we need benchmarks to find this cut-off, if it even makes sense to have logic like this.
This is determined by how much faster the normal `aggregate_verify()` is compared to the multiplication loop of `GTElement`s.

Reply:
I think you should still make this benchmark to know if there's still a use for a heuristic like this